File: src/api/git.ts
import { invoke } from "@tauri-apps/api/core";
import type { GitInfo } from "@/types";
import { ApiError } from "./project";

/**
 * Git 分析 API 封装（对齐后端 commands::git::get_project_git_info）。
 *
 * 后端签名：`get_project_git_info(projectId) -> Result<Option<GitInfo>, String>`
 * - 项目不存在 → Err
 * - 非 git 仓库 → null（优雅降级，GitInfo.is_git_repo 为 false）
 * - git 仓库 → GitInfo
 *
 * ⚠️ 只读：后端仅用 git2 只读 API，绝不修改 git 状态。
 */

async function call<T>(fn: () => Promise<T>): Promise<T> {
  try {
    return await fn();
  } catch (e) {
    throw new ApiError(typeof e === "string" ? e : e instanceof Error ? e.message : String(e));
  }
}

/** 获取指定项目的 git 信息（只读）；非 git 仓库返回 null */
export function getProjectGitInfo(projectId: number): Promise<GitInfo | null> {
  return call(() =>
    invoke<GitInfo | null>("get_project_git_info", { projectId })
  );
}

File: src-tauri/src/commands/git.rs
//! Git 相关命令（薄壳层）。
//!
//! 职责：`project_id` → 从库取项目路径 → 调 `core::git` 只读分析 → 返回 `GitInfo`。
//! 本层仅做适配，不含业务逻辑。

use std::sync::Mutex;

use tauri::State;

use crate::core::database::Database;
use crate::core::git;
use crate::core::models::GitInfo;

/// 获取指定项目的 git 信息（只读）。
///
/// - `project_id` 对应的项目不存在 → 返回 `Err("项目不存在: ...")`。
/// - 项目非 git 仓库 / `.git` 损坏 / 权限不足 → 返回 `Ok(None)`（优雅降级，
///   `GitInfo.is_git_repo` 为 `false`）。
///
/// 返回 `Result<Option<GitInfo>, String>`，沿用现有错误格式。
#[tauri::command]
pub fn get_project_git_info(
    db: State<'_, Mutex<Database>>,
    project_id: i64,
) -> Result<Option<GitInfo>, String> {
    let db = db.lock().map_err(|e| e.to_string())?;
    let detail = db
        .get_project_detail(project_id)
        .map_err(|e| format!("查询项目失败: {e}"))?
        .ok_or_else(|| format!("项目不存在: {project_id}"))?;

    let info = git::analyze_git(std::path::Path::new(&detail.path));

    // 非 git 仓库：优雅降级返回 None
    if !info.is_git_repo {
        return Ok(None);
    }
    Ok(Some(info))
}
